### ISSUE TITLE
Top level nulls for StructArrays are not handled appropriately

### ISSUE THREAD

**Issue by theelderbeever** [chmp/serde_arrow#90]
For serde types where a field is an `Option<ChildStruct>` and the read in values for a `Vec<ParentStruct>` all have `null` for the `parent.child` field the correct type Field value is not intuited.

In the example below the resultant RecordBatch has the schema of 

```
RecordBatch {
    schema: Schema {
        fields: [
            Field {
                name: "distribution",
                data_type: Null,
                nullable: true,
                dict_id: 0,
                dict_is_ordered: false,
                metadata: {},
            },
        ],
        metadata: {},
    },
    columns: [
        NullArray(4),
    ],
    row_count: 4,
}
```

while it should still understand that there is a `StructArray` with null values.

```rust
#[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
pub struct Distribution {
    pub samples: Vec<f64>,
    pub statistic: String,
}

#[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
pub struct VectorMetric {
    pub distribution: Option<Distribution>,
}

let json = "[{}, {}, {}]";

let metrics = serde_json::from_str(json).unwrap();

let fields = serde_arrow::arrow::serialize_into_fields(
    &metrics,
    TracingOptions::default().allow_null_fields(true),
)
.unwrap();
let arrays = serde_arrow::arrow::serialize_into_arrays(&fields, &metrics).unwrap();

let batch = RecordBatch::try_new(Arc::new(Schema::new(fields)), arrays).unwrap();
println!("{:#?}", batch);
```

**Comment by chmp** [comment]
Hey @theelderbeever 

Thanks for reporting this issue. Unfortunatenly, the behavior you observed is unavoidable by using serde to determine the schema. serde only tells you about values that are actually there. Therefore if all values are null, there is no way to detect the schema information.

I tried to spell out the requirements in the [docs](https://docs.rs/serde_arrow/latest/serde_arrow/arrow/fn.serialize_into_fields.html)

> To correctly record the type information make sure to:
>
> - include values for Option<T>
> - include all variants of an enum
> - include at least single element of a list or a map

Maybe there is a better way to phrase it ior make it a more prominent warning. What do you think?

**Comment by theelderbeever** [comment]
@chmp Ah yeah this one is mostly on me for not quite fully digesting the requirements. I think it could be valuable to just pull attention to it a little more and be more overt? Maybe something like the following?

> **NOTE**: Tracing **only** uses the data dynamically available to it to generate the schema.  In order to fully resolve the correct schema from Rust objects the following must be available in the data being traced.
> - Both `None` and `Some(T)` for `Option<T>` fields
> - All variants of enum fields
> - At least one element of a list field
> - At least one example of map types with all possible keys

I think the json schema would probably be the best for my situation honestly. I am probably going to store the schema in a postgres database or something that can be fetched by different datafusion instances.

Out of curiosity since I don't fully have a grasp on the macro system yet but, what do you think of using something like the following? I am not sure the value it brings but it would somewhat mimic the serde api I guess?
```rust
pub struct MyStruct {
  #[arrow("Int64", nullable = true)]
  integer_field: i64,
  #[arrow("Utf8", nullable = true)]
  string_field: String,
  #[arrow("Map(key = "Utf8", value = "Uft8"), nullable = true)]
  map_field: HashMap(String, String),
}
```

**Comment by chmp** [comment]
Thanks. I will spend some time overhauling the docs the coming days and try to clarify it. Just FYI: For options the None case is not required, only the Some case.

Regardin your macro idea: while I like it, I suspect it would take quite some effort and also be quite a departure from the current way `serde_arrow` works. Something for me to think about :)

**Comment by chmp** [comment]
I did some research and stumbled over [serde-reflection](https://docs.rs/serde-reflection/latest/serde_reflection/). After some time of wrapping my head around of how it works, I now have a working prototype that uses deserialization to trace the schema, even without any examples. 

For your struct, it works as expected:

```rust
#[test]
fn issue_90() {
    use {GenericDataType as T, GenericField as F};

    #[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
    pub struct Distribution {
        pub samples: Vec<f64>,
        pub statistic: String,
    }

    #[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
    pub struct VectorMetric {
        pub distribution: Option<Distribution>,
    }

    // the new tracing call (will not be exposed to end users)
    let actual = trace::<VectorMetric>("root").unwrap();
    
    let expected = F::new("root", T::Struct, false).with_child(
        F::new("distribution", T::Struct, true)
            .with_child(F::new("samples", T::List, false).with_child(F::new("item", T::F64, false)))
            .with_child(F::new("statistic", T::Utf8, false)),
    );

    assert_eq!(actual, expected);
}
```

Once I find some more time, I will polish this protoype a bit and integrate it with the rest of `serde_arrow`. Let's see how it handles the various test cases. Hopefully, I can push a pre-release with the updated docs and a preview of this tracing approach in the coming days, .

**Comment by theelderbeever** [comment]
Nice find! I am going to try and wrap my head around it too. Any idea whether it could support the arrow `Map` type too?

**Comment by chmp** [comment]
Both methods will support map types. 

At the moment `serialize_into_fields` does already support Arrow maps (with `TracingOptions::default().map_as_struct(false)`, as does the JSON encodoing. See the tests [here](https://github.com/chmp/serde_arrow/blob/main/serde_arrow/src/test_impls/map.rs), if you're interested what is tested.

**Comment by chmp** [comment]
I published a new pre-release `serde_arrow=0.9.0-rc.1`, that includes the new type tracing support. There are some minor breaking changes and some deprecations, as I simplified the API a bit. See the docs [here](https://docs.rs/serde_arrow/0.9.0-rc.1/serde_arrow/).

The new feature should [simplify your use case quite a bit](https://github.com/chmp/serde_arrow/blob/1f0242f4accc2f13690ea428947471cf5642bc57/serde_arrow/src/test_end_to_end/issue_90.rs#L25).

**Comment by theelderbeever** [comment]
@chmp Nice! Those changes are looking really good. The samples docs are quite clear now too. Thanks for the responsiveness and quick updates!
